### ISSUE TITLE
pgAdmin4 with AAD-SSO on Kubernetes - "The CSRF tokens do not match"

### ISSUE THREAD

**Issue by AlexanderInova** [pgadmin-org/pgadmin4#5585]
**TL;DR**

pgAdmin4 with AAD-SSO on Kubernetes behind an istio-ingress-controller throws HTTP-400 "The CSRF tokens do not match" errors every now and then, and thus stopping the current pgAdmin4-session from working. 

---

**Full description**

We run pgAdmin4 in our Kubernetes cluster. It is accessed via istio-ingress-controller, which in turn is accessed via a layer 4 (Azure) load balancer. Authentication is done via SSO (Azure Active Directory). Since the fix in the previous version, the sign-in finally works properly without volatile errors - many thanks for that! 

Unfortunately, another bug continues to plague us: Our running session is interrupted practically at random. The server log outputs 
```
ERROR	pgadmin:	400 Bad Request: The CSRF tokens do not match.
Traceback (most recent call last):
  File "/venv/lib/python3.10/site-packages/flask_wtf/csrf.py", line 261, in protect
    validate_csrf(self._get_csrf_token())
  File "/venv/lib/python3.10/site-packages/flask_wtf/csrf.py", line 115, in validate_csrf
    raise ValidationError("The CSRF tokens do not match.")
wtforms.validators.ValidationError: The CSRF tokens do not match.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.10/site-packages/flask/app.py", line 1515, in full_dispatch_request
    rv = self.preprocess_request()
  File "/venv/lib/python3.10/site-packages/flask/app.py", line 1857, in preprocess_request
    rv = self.ensure_sync(before_func)()
  File "/venv/lib/python3.10/site-packages/flask_wtf/csrf.py", line 229, in csrf_protect
    self.protect()
  File "/venv/lib/python3.10/site-packages/flask_wtf/csrf.py", line 264, in protect
    self._error_response(e.args[0])
  File "/venv/lib/python3.10/site-packages/flask_wtf/csrf.py", line 307, in _error_response
    raise CSRFError(reason)
flask_wtf.csrf.CSRFError: 400 Bad Request: The CSRF tokens do not match.
``` 

For us, there is no predictable pattern behind this. Sometimes you can work through the whole day without the problem occurring. But most of the times you can't. Often you don't even manage to connect to the first server before the interface switches to "loading infidelity". In the network monitor you can see that HTTP-400 errors are thrown. 
Currently our employees are thrown out of the session appr. 5-50 times a day on average. Logically, it is not possible to save the current work when the issue occurs. 

Is anyone aware of the problem and a solution to it? 
If not: What information can we provide in detail to track down the problem?

**Comment by yogeshmahajan-1903** [comment]
@Alexanerinova 
Are you using 'LoadBalancer' type istio-ingressgateway?
LoadBalancer will not work as  pgAdmin maintains a pool of database connections within each instance, and neither this or session data can be shared between multiple instances.

**Comment by Alexanerinova** [comment]
Hello @yogeshmahajan-1903 , 

thank you for your response. Our pgAdmin4 deployment is configured to run only one pod/instance at a time. 

pgAdmin4 works fine except for the seemingly random CSRF token errors. 

I drew a little graphic to help understanding the communication path: 
![Untitled (1)](https://user-images.githubusercontent.com/113263664/204476358-7c900dd3-d4df-4baf-9fd0-802d79e11fd2.png)

Maybe I should mention that a pgBouncer instance is running on the Postgres server, even I don't think postgres itself is related to this issue at all, since the CSRF token is most probably handled by pgAdmin4 itself. It's not the database connection that is dropped, but the authentication from client browser to pgAdmin4. The sign-in session is dead after that. Since pgAdmin4 doesn't handle broken sessions at all, this behavior renders pgAdmin4 completely unusable and unresponsive until you reload the site.

I apologize for my probably broken English.

**Comment by yogeshmahajan-1903** [comment]
@Alexanerinova 
Would you mind sharing manifesto file for above setup?

**Comment by Alexanerinova** [comment]
@yogeshmahajan-1903 Of course, no problem: [pgadmin-manifest-snapshot.yaml.txt](https://github.com/pgadmin-org/pgadmin4/files/10276638/pgadmin-manifest-snapshot.yaml.txt)

**Comment by Alexanerinova** [comment]
I found a workaround, actually within the pgAdmin4 documentation. This issue seems to be related to the `ENHANCED_COOKIE_PROTECTION` feature. Setting this configuration value to `False` seems to avoid the issue. 

See #2843 for further details.

@yogeshmahajan-1903 Is there a known way to allow `ENHANCED_COOKIE_PROTECTION` behind a load balancer?

**Comment by adityatoshniwal** [comment]
Closing this as there is a solution.
